---
title: "Preprocessing"
author: "Antone Jung"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center")
```

#Goals

1. Trim raw reads
2. Detect strandedness

#Input

1. Raw `.fastq.gz` files from Michelle's Diurnal BARI06 series, Byron's Organ Atlas BARI06 data
2. M82/BARI06 Transcriptome

#Output

1. Trimmed `.fastq.gz` files
2. Salmon detection files

```{r set-seed}
set.seed(10202005)
```

```{bash fastp-trimming-egg, eval = FALSE}
egg_michlle_trim () {
  base1=$(basename "$1" _001.fastq.gz)
  base2=${base1/R1/R2}
	r2_filename=${1/R1/R2}
	/programs/fastp-0.23.4/fastp -i "$1" -I "$r2_filename" -o data/01_preprocessing/egg_trimmed_reads/${base1}"_trimmed.fastq.gz" -O data/01_preprocessing/egg_trimmed_reads/${base2}"_trimmed.fastq.gz" -h data/01_preprocessing/egg_fastp_reports/byron/${base1}.html -j /dev/null --thread 16
}
export -f egg_michelle_trim
parallel egg_michelle_trim ::: data/00_raw_input_reads/egg/michelle/*R1.fastq.gz

egg_byron_trim () {
  base1=$(basename "$1" .fastq.gz)
  base2=${base1/R1/R2}
	r2_filename=${1/R1/R2}
	/programs/fastp-0.23.4/fastp -i "$1" -I "$r2_filename" -o data/01_preprocessing/egg_trimmed_reads/${base1}"_trimmed.fastq.gz" -O data/01_preprocessing/egg_trimmed_reads/${base2}"_trimmed.fastq.gz" -h data/01_preprocessing/egg_fastp_reports/byron/${base1}.html -j /dev/null --thread 16
}
export -f egg_byron_trim
parallel egg_byron_trim ::: data/00_raw_input_reads/egg/byron/*R1.fastq.gz
```

```{bash fastp-trimming-tom, eval = FALSE}
tom_michelle_trim () {
  base1=$(basename "$1" _001.fastq.gz)
  base2=${base1/R1/R2}
	r2_filename=${1/R1/R2}
	/programs/fastp-0.23.4/fastp -i "$1" -I "$r2_filename" -o data/01_preprocessing/tom_trimmed_reads/${base1}"_trimmed.fastq.gz" -O data/01_preprocessing/tom_trimmed_reads/${base2}"_trimmed.fastq.gz" -h data/01_preprocessing/tom_fastp_reports/michelle/${base1}.html -j /dev/null --thread 24
}
export -f tom_michelle_trim
parallel tom_michelle_trim ::: data/00_raw_input_reads/tom/michelle/*R1_001.fastq.gz
```

```{bash salmon-strandedness-egg, eval = FALSE}
/programs/salmon-1.10.0/bin/salmon index -t resources/eggplant_transcriptome.fasta -i resources/egg_salmon_index

egg_salmon () {
  base=$(basename $1 _R1_trimmed.fastq.gz)
	r2_filename=${1/R1/R2}
	/programs/salmon-1.10.0/bin/salmon quant -l A -1 $1 -2 $r2_filename -i resources/egg_salmon_index --validateMappings -o data/01_preprocessing/egg_strand_detection/${base}_quant
}
export -f egg_salmon
parallel egg_salmon ::: data/01_preprocessing/egg_trimmed_reads/*R1*
```

# Michelle eggplant dataset Automatically detected most likely library type as ISF

# Byron eggplant dataset Automatically detected most likely library type as ISR

```{bash salmon-strandedness-tom, eval = FALSE}
/programs/salmon-1.10.0/bin/salmon index -t resources/tomato_transcriptome.fasta -i resources/tom_salmon_index

tom_salmon () {
  base=$(basename $1 _R1_trimmed.fastq.gz)
	r2_filename=${1/R1/R2}
	/programs/salmon-1.10.0/bin/salmon quant -l A -1 $1 -2 $r2_filename -i resources/tom_salmon_index --validateMappings -o data/01_preprocessing/tom_strand_detection/${base}_quant
}
export -f tom_salmon
parallel tom_salmon ::: data/01_preprocessing/tom_trimmed_reads/*R1*
```

# Michelle tomato dataset Automatically detected most likely library type as ISF

# Dr. Frank tomato dataset 