---
title: "Classifying transcripts"
author: "Antone Jung"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center")
```

#Goals

1. Map transcripts using Hisat2
2. Assemble transcripts using Stringtie
3. Compare transcripts using GffCompare

#Inputs

1. Trimmed `fastq.gz` files
2. `.gtf` files from priorly annotated genomes (Nat)
3. Salmon-detected strandedness from Nat (Michelle M82 and BARI06 = FR,
Byron BARI06 = RF)
4. Hisat2 indices `.ht2` files

#Outputs

1. Transcripts tagged as antisense, sense, or chr00mapping

```{r set-seed}
set.seed(10202005)
```

```{bash set-seed}
SEED=10202005
```

```{bash hisat-mapping}
source /programs/HISAT2/hisat2.sh

for r1_filename in data/01_preprocessing/trimmed_reads/egg/EE*R1_trimmed.fastq.gz
	do
	base=$(basename $r1_filename _R1_trimmed.fastq.gz)
	r2_filename=${r1_filename/R1/R2}
	hisat2 -p 20 -x resources/egg_hisat_indices/nategg_index --max-intronlen 1500 --dta-cufflinks --rna-strandness FR -1 $r1_filename -2 $r2_filename | samtools sort -o data/02_transcripts/egg_01mapped/${base}.bam
	done
	

hisat_RF () {
  base=$(basename "$1" _R1_trimmed.fastq.gz)
	r2_filename=${1/R1/R2}
	hisat2 -p 20 -x resources/egg_hisat_indices/nategg_index --max-intronlen 1500 --dta-cufflinks --rna-strandness RF -1 $1 -2 $r2_filename | samtools sort -o data/02_transcripts/egg_01mapped/${base}.bam
}
export -f hisat_RF
parallel hisat_RF ::: data/01_preprocessing/trimmed_reads/egg/123*R1*
```

```{bash string-tie-assembly}
export PATH=/programs/stringtie-2.2.1:$PATH

ls data/02_transcripts/egg_01mapped/*.bam | parallel samtools index '{}'

stringtie_assemble_FR () {
  base=$(basename $1 .bam)
  stringtie -p 8 $1 -G resources/agat_bari06.gtf -o data/02_transcripts/egg_02assembled/${base}_stringtie.gtf -f 0.05 -j 5 -c 5 -s 15 --fr
}
export -f stringtie_assemble_FR

parallel stringtie_assemble_FR ::: data/02_transcripts/egg_01mapped/EE*.bam

stringtie_assemble_RF () {
  base=$(basename $1 .bam)
  stringtie -p 8 $1 -G resources/agat_bari06.gtf -o data/02_transcripts/egg_02assembled/${base}_stringtie.gtf -f 0.05 -j 5 -c 5 -s 15 --rf
}
export -f stringtie_assemble_RF

parallel stringtie_assemble_RF ::: data/02_transcripts/egg_01mapped/123*.bam
```

```{bash merge}
ls -1 data/02_transcripts/egg_02assembled/*.gtf > data/02_transcripts/egg_03merged/egg_mergelist.txt

stringtie -p 8 --merge -G resources/agat_bari06.gtf -m 200 -F 0.5 -o data/02_transcripts/egg_03merged/egg_merged.gtf data/02_transcripts/egg_03merged/egg_mergelist.txt
# minimum 200 bp and minimum FPKM (Wang et al., 2015)

awk '$7 != "."' data/02_transcripts/egg_03merged/egg_merged.gtf > data/02_transcripts/egg_03merged/egg_no_unstranded_merged.gtf
```

```{bash classify-transcripts}
gffcompare -r resources/agat_bari06.gtf -o data/02_transcripts/egg_04gff_compare/egg_gffcompare data/02_transcripts/egg_no_unstranded_merged.gtf

grep 'ch00' data/02_transcripts/egg_04gffcompare/egg_gffcompare.annotated.gtf > data/02_transcripts/egg_05classified/egg_ch00.gtf

grep -v "ch00" data/02_transcripts/egg_04gffcompare/egg_gffcompare.annotated.gtf | grep 'class_code "u"' > data/02_transcripts/egg_05classified/egg_intergenic.gtf

grep -v "ch00" data/02_transcripts/egg_04gffcompare/egg_gffcompare.annotated.gtf | grep 'class_code "x"' > data/02_transcripts/egg_05classified/egg_antisense.gtf

grep -v "ch00" data/02_transcripts/egg_04gffcompare/egg_gffcompare.annotated.gtf | grep -E 'class_code "u"|class_code "x"' > data/02_transcripts/egg_05classified/egg_antint.gtf

# u = intergenic, x = antisense
```

```{bash gffread-gtf-to-fa}
/programs/gffread-0.9.12/gffread/gffread data/02_transcripts/egg_05classified/egg_antint.gtf -g resources/galaxy_bari06.fa -w data/02_transcripts/egg_05classified/egg_antint.fasta
```