---
title: "Filtering for high confidence lncRNAs"
author: "Antone Jung"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center")
```

#Goals

1. Annotate and filter with CPC2, TransDecoder, Infernal

#Inputs

1. Intergenic and antisense RNAs

#Outputs

1. Fully annotated gtf with the results of each filtering step

```{r set-seed}
set.seed(10202005)
```

```{bash set-seed}
SEED=10202005
```

```{bash CPC2}
resources/CPC2_standalone-1.0.1/bin/CPC2.py -i data/02_transcripts/egg_05classified/egg_antint.fasta -o data/03_filtering/egg_01CPC2/egg_antintCPC2

grep "noncoding" data/03_filtering/egg_01CPC2/egg_antintCPC2.txt | grep -o 'MSTRG.[0/-9]*.[0/-9]*' | sed 's/$/\"/' | sed 's/^/\"/' > data/03_filtering/egg_01CPC2/egg_noncoding.txt

grep -w "coding" data/03_filtering/egg_01CPC2/egg_antintCPC2.txt | grep -o 'MSTRG.[0/-9]*.[0/-9]*' | sed 's/$/\"/' | sed 's/^/\"/' > data/03_filtering/egg_01CPC2/egg_coding.txt

grep -h -f data/03_filtering/egg_01CPC2/egg_noncoding.txt data/02_transcripts/egg_05classified/egg_antint.gtf | sed 's/$/ CPC2 \"noncoding\";/' > data/03_filtering/egg_01CPC2/egg_CPC2_annotated.gtf

grep -h -f data/03_filtering/egg_01CPC2/egg_coding.txt data/02_transcripts/egg_05classified/egg_antint.gtf | sed 's/$/ CPC2 \"coding\";/' >> data/03_filtering/egg_01CPC2/egg_CPC2_annotated.gtf

sort data/03_filtering/egg_01CPC2/egg_CPC2_annotated.gtf -o data/03_filtering/egg_01CPC2/egg_CPC2_annotated.gtf
```

```{bash transdecoder}
cd data/03_filtering/egg_02TransDecoder

/programs/TransDecoder-v5.5.0/TransDecoder.LongOrfs -t ../../02_transcripts/egg_05classified/egg_antint.fasta

/programs/hmmer/bin/hmmsearch --cpu 16 -E 1e-10 --domtblout egg_pfam.domtblout ../../../resources/Pfam-A.hmm egg_antint.fasta.transdecoder_dir/longest_orfs.pep

cd ../../../resources/swissprot
blastp -query ../../data/03_filtering/egg_02TransDecoder/egg_antint.fasta.transdecoder_dir/longest_orfs.pep -db swissprot -max_target_seqs 1 -outfmt 6 -evalue 1e-5 -out ../../data/03_filtering/egg_02TransDecoder/egg_blastp.outfmt6

cd ../../data/03_filtering/egg_02TransDecoder
/programs/TransDecoder-v5.5.0/TransDecoder.Predict -t ../../02_transcripts/egg_05classified/egg_antint.fasta --retain_pfam_hits egg_pfam.domtblout --retain_blastp_hits egg_blastp.outfmt6 

grep -o '>MSTRG\S*' egg_antint.fasta.transdecoder.cds | grep -o 'MSTRG\.[0/-9]*\.[0/-9]*' | sed 's/$/\"/' | sed 's/^/\"/' | uniq > egg_ORFs.txt

grep -h -v -f egg_ORFs.txt ../egg_01CPC2/egg_CPC2_annotated.gtf | grep -o 'MSTRG\.[0/-9]*\.[0/-9]*' | sed 's/$/\"/' | sed 's/^/\"/' > egg_nonORFs.txt

cd ../../..

grep -h -f data/03_filtering/egg_02TransDecoder/egg_nonORFs.txt data/03_filtering/egg_01CPC2/egg_CPC2_annotated.gtf | sed 's/$/ TransDecoder \"nonORFs\";/' > data/03_filtering/egg_02TransDecoder/egg_TransDecoder_annotated.gtf

grep -h -f data/03_filtering/egg_02TransDecoder/egg_ORFs.txt data/03_filtering/egg_01CPC2/egg_CPC2_annotated.gtf | sed 's/$/ TransDecoder \"ORFs\";/' >> data/03_filtering/egg_02TransDecoder/egg_TransDecoder_annotated.gtf

sort data/03_filtering/egg_02TransDecoder/egg_TransDecoder_annotated.gtf -o data/03_filtering/egg_02TransDecoder/egg_TransDecoder_annotated.gtf
```

```{bash infernal}
export PATH=/programs/infernal-1.1.3/bin:$PATH
cd resources/rfam

cmscan -o ../../data/03_filtering/egg_03RFAM/infernal.log --noali --tblout ../../data/03_filtering/egg_03RFAM/egg_rfam.tblout --rfam --cpu 24 Rfam.cm ../../data/02_transcripts/egg_05classified/egg_antint.fasta
cd ../..

grep '!' data/03_filtering/egg_03RFAM/egg_rfam.tblout > data/03_filtering/egg_03RFAM/egg_rfam_sig.txt

grep -o 'MSTRG\.[0/-9]*\.[0/-9]*' data/03_filtering/egg_03RFAM/egg_rfam_sig.txt | uniq | sed 's/$/\"/' | sed 's/^/\"/' > data/03_filtering/egg_03RFAM/egg_knownRNA.txt

grep -h -v -f data/03_filtering/egg_03RFAM/egg_knownRNA.txt data/03_filtering/egg_02TransDecoder/egg_TransDecoder_annotated.gtf | grep -o 'MSTRG\.[0/-9]*\.[0/-9]*' | sed 's/$/\"/' | sed 's/^/\"/' > data/03_filtering/egg_03RFAM/egg_not_knownRNA.txt

grep -h -f data/03_filtering/egg_03RFAM/egg_knownRNA.txt data/03_filtering/egg_02TransDecoder/egg_TransDecoder_annotated.gtf | sed 's/$/ RFAM \"known\";/' > data/03_filtering/egg_03RFAM/egg_RFAM_annotated.gtf

grep -h -f data/03_filtering/egg_03RFAM/egg_not_knownRNA.txt data/03_filtering/egg_02TransDecoder/egg_TransDecoder_annotated.gtf | sed 's/$/ RFAM \"not known\";/' >> data/03_filtering/egg_03RFAM/egg_RFAM_annotated.gtf

sort data/03_filtering/egg_03RFAM/egg_RFAM_annotated.gtf -o data/03_filtering/egg_03RFAM/egg_RFAM_annotated.gtf
```

```{bash feature-counts}
#transcript level
/programs/subread-2.1.1/bin/featureCounts -T 24 -p -a data/02_transcripts/egg_05classified/egg_antint.gtf -F 'GTF' -g 'transcript_id' -t transcript -o data/03_filtering/egg_04FeatureCounts/egg_transcript_counts.txt data/02_transcripts/egg_01mapped/*.bam

#gene level
/programs/subread-2.1.1/bin/featureCounts -T 24 -p -a data/02_transcripts/egg_05classified/egg_antint.gtf -F 'GTF' -t transcript -o data/03_filtering/egg_04FeatureCounts/egg_gene_counts.txt data/02_transcripts/egg_01mapped/*.bam

grep 'MSTRG' data/03_filtering/egg_04FeatureCounts/egg_transcript_counts.txt > data/03_filtering/egg_04FeatureCounts/egg_transcript_counts_no_metadata.txt

grep 'MSTRG' data/03_filtering/egg_04FeatureCounts/egg_gene_counts.txt > data/03_filtering/egg_04FeatureCounts/egg_gene_counts_no_metadata.txt
```

```{r filter-low-expression}
library(dplyr)
egg_gene_counts <- read.csv("data/03_filtering/egg_04FeatureCounts/egg_gene_counts.txt", sep = "\t", header = FALSE, skip = 2)
egg_transcript_counts <- read.csv("data/03_filtering/egg_04FeatureCounts/egg_transcript_counts.txt", sep = "\t", header = FALSE, skip = 2) 

expression_level <- function(min_reads, min_count, arr){
  sum_reads <- 0
  for (num in arr){
    if (num >= min_count){
      sum_reads <- sum_reads + 1
    }
  }
  return (sum_reads >= min_reads)
}

gene_rows_to_keep <-c()
transcript_rows_to_keep <- c()

for (r in 1:nrow(egg_gene_counts)){
  if(expression_level(3,3,egg_gene_counts[r,7:108])){
    gene_rows_to_keep<- append(gene_rows_to_keep,r)
  }
}
for (r in 1:nrow(egg_transcript_counts)){
  if(expression_level(3,3,egg_transcript_counts[r,7:108])){
    transcript_rows_to_keep<- append(transcript_rows_to_keep,r)
  }
}

egg_gene_counts_filtered <- egg_gene_counts[gene_rows_to_keep,]
egg_transcript_counts_filtered <- egg_transcript_counts[transcript_rows_to_keep,]

egg_gene_counts_discard <- egg_gene_counts[-(gene_rows_to_keep),]
egg_transcript_counts_discard <- egg_transcript_counts[-(transcript_rows_to_keep),]

write.table(egg_gene_counts_filtered$V1, file = "data/03_filtering/egg_04FeatureCounts/egg_gene_highexpr.txt", col.names = FALSE, row.names=FALSE, sep="\t", quote = FALSE)
write.table(egg_transcript_counts_filtered$V1, file = "data/03_filtering/egg_04FeatureCounts/egg_transcript_highexpr.txt", col.names = FALSE, row.names=FALSE, sep="\t", quote = FALSE)

write.table(egg_gene_counts_discard$V1, file = "data/03_filtering/egg_04FeatureCounts/egg_gene_lowexpr.txt", col.names = FALSE, row.names=FALSE, sep="\t", quote = FALSE)
write.table(egg_transcript_counts_discard$V1, file = "data/03_filtering/egg_04FeatureCounts/egg_transcript_lowexpr.txt", col.names = FALSE, row.names=FALSE, sep="\t", quote = FALSE)
```

```{bash annotate-with-expr}
sed -i 's/$/\"/' data/03_filtering/egg_04FeatureCounts/egg_transcript_lowexpr.txt
sed -i 's/^/\"/' data/03_filtering/egg_04FeatureCounts/egg_transcript_lowexpr.txt

sed -i 's/$/\"/' data/03_filtering/egg_04FeatureCounts/egg_transcript_highexpr.txt
sed -i 's/^/\"/' data/03_filtering/egg_04FeatureCounts/egg_transcript_highexpr.txt

grep -h -f data/03_filtering/egg_04FeatureCounts/egg_gene_highexpr.txt data/03_filtering/egg_03RFAM/egg_RFAM_annotated.gtf | sed 's/$/ Gene3in3 \"high\";/' > data/03_filtering/egg_04FeatureCounts/egg_geneexpr_annotated.gtf

grep -h -f data/03_filtering/egg_04FeatureCounts/egg_gene_lowexpr.txt data/03_filtering/egg_03RFAM/egg_RFAM_annotated.gtf | sed 's/$/ Gene3in3 \"low\";/' >> data/03_filtering/egg_04FeatureCounts/egg_geneexpr_annotated.gtf

grep -h -f data/03_filtering/egg_04FeatureCounts/egg_transcript_highexpr.txt data/03_filtering/egg_04FeatureCounts/egg_geneexpr_annotated.gtf | sed 's/$/ Transcript3in3 \"high\";/' > data/03_filtering/egg_04FeatureCounts/egg_expr_annotated.gtf

grep -h -f data/03_filtering/egg_04FeatureCounts/egg_transcript_lowexpr.txt data/03_filtering/egg_04FeatureCounts/egg_geneexpr_annotated.gtf | sed 's/$/ Transcript3in3 \"low\";/' >> data/03_filtering/egg_04FeatureCounts/egg_expr_annotated.gtf

sort data/03_filtering/egg_04FeatureCounts/egg_expr_annotated.gtf -o data/03_filtering/egg_04FeatureCounts/egg_expr_annotated.gtf
```

```{bash blastn-arabidopsis-r-trna}
cd resources/at_trna_rrna.BLAST
blastn -query ../../data/02_transcripts/egg_05classified/egg_antint.fasta -db at_trna_rrna -max_target_seqs 1 -outfmt 6 -evalue 1e-5 -out ../../data/03_filtering/egg_05blastn/egg_at_blastn.outfmt6

cd ../solanum.lnc.BLAST
blastn -query ../../data/02_transcripts/egg_05classified/egg_antint.fasta -db solanum.plncdb -max_target_seqs 1 -outfmt 6 -evalue 1e-5 -out ../../data/03_filtering/egg_05blastn/egg_sol_blastn.outfmt6

grep -o 'MSTRG\.[0/-9]*\.[0/-9]*' data/03_filtering/egg_05blastn/egg_at_blastn.outfmt6 | uniq | sed 's/$/\"/' | sed 's/^/\"/' > data/03_filtering/egg_05blastn/egg_at_trna_rrna.txt

grep -o 'MSTRG\.[0/-9]*\.[0/-9]*' data/03_filtering/egg_05blastn/egg_sol_blastn.outfmt6 | uniq | sed 's/$/\"/' | sed 's/^/\"/' > data/03_filtering/egg_05blastn/egg_sol_lncRNA.txt

grep -h -f data/03_filtering/egg_05blastn/egg_at_trna_rrna.txt data/03_filtering/egg_04FeatureCounts/egg_expr_annotated.gtf | sed 's/$/ Athal blast \"trna or rrna\";/' > data/03_filtering/egg_05blastn/egg_at_annotated.gtf

grep -v -h -f data/03_filtering/egg_05blastn/egg_at_trna_rrna.txt data/03_filtering/egg_04FeatureCounts/egg_expr_annotated.gtf | sed 's/$/ Athal blast \"no hit\";/' >> data/03_filtering/egg_05blastn/egg_at_annotated.gtf

grep -h -f data/03_filtering/egg_05blastn/egg_sol_lncRNA.txt data/03_filtering/egg_05blastn/egg_at_annotated.gtf | sed 's/$/ Solanum blast \"known lncRNA\";/' > data/03_filtering/egg_05blastn/egg_sol_annotated.gtf

grep -v -h -f data/03_filtering/egg_05blastn/egg_sol_lncRNA.txt data/03_filtering/egg_05blastn/egg_at_annotated.gtf | sed 's/$/ Solanum blast \"no hit\";/' >> data/03_filtering/egg_05blastn/egg_sol_annotated.gtf

sort data/03_filtering/egg_05blastn/egg_sol_annotated.gtf -o data/03_filtering/egg_05blastn/egg_sol_annotated.gtf
```

```{bash filter-for-lncRNA}
grep "CPC2 \"noncoding\"\|Solanum blast \"known lncRNA\"" data/03_filtering/egg_05blastn/egg_sol_annotated.gtf | grep "TransDecoder \"nonORFs\"\|Solanum blast \"known lncRNA\"" | grep "RFAM \"not known\"\|Solanum blast \"known lncRNA\"" | grep "Gene3in3 \"high\"\|Solanum blast \"known lncRNA\"" | grep "Athal blast \"no hit\"\|Solanum blast \"known lncRNA\"" > data/03_filtering/egg_06lncRNA/egg_lncRNA_transcripts.gtf

grep "\"MSTRG\.[0\-9]*\.1\"" data/02_transcripts/egg_05classified/egg_antint.gtf > data/02_transcripts/egg_05classified/egg_antint_genes.gtf

grep "\"MSTRG\.[0\-9]*\.1\"" data/03_filtering/egg_06lncRNA/egg_lncRNA_transcripts.gtf > data/03_filtering/egg_06lncRNA/egg_lncRNA_genes.gtf

grep "Solanum blast \"known lncRNA\"" data/03_filtering/egg_06lncRNA/egg_lncRNA_transcripts.gtf | grep "CPC2 \"noncoding\"" | grep "TransDecoder \"nonORFs\"" | grep "RFAM \"not known\"" | grep "Gene3in3 \"high\"" | grep "Athal blast \"no hit\"" | wc -l
```