---
title: "Filtering for high confidence lncRNAs"
author: "Antone Jung"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center")
```

#Goals

1. Annotate and filter with CPC2, TransDecoder, Infernal

#Inputs

1. Intergenic and antisense RNAs

#Outputs

1. lncRNAs

```{r set-seed}
set.seed(10202005)
```

```{bash set-seed}
SEED=10202005
```

```{bash CPC2}
resources/CPC2_standalone-1.0.1/bin/CPC2.py -i data/02_transcripts/egg_05classified/egg_antint.fasta -o data/03_filtering/egg_01CPC2/egg_antintCPC2

grep "noncoding" data/03_filtering/egg_01CPC2/egg_antint_CPC2.txt > data/03_filtering/egg_01CPC2/egg_noncoding.txt

grep "coding" data/03_filtering/egg_01CPC2/egg_antint_CPC2.txt > data/03_filtering/egg_01CPC2/egg_coding.txt

grep -o '\S*MSTRG.\S*' data/03_filtering/egg_01CPC2/egg_noncoding.txt > data/03_filtering/egg_01CPC2/egg_noncoding_names.txt

grep -h -f data/03_filtering/egg_01CPC2/egg_noncoding_names.txt data/02_transcripts/egg_05classified/egg_antint.gtf > data/03_filtering/egg_01CPC2/egg_noncoding.gtf

grep -o "StringTie" data/03_filtering/egg_01CPC2/egg_noncoding.gtf | wc -l

/programs/gffread-0.9.12/gffread/gffread data/03_filtering/egg_01CPC2/egg_noncoding.gtf -g resources/galaxy_bari06.fa -w data/03_filtering/egg_01CPC2/egg_noncoding.fasta
```

```{bash transdecoder}
cd data/03_filtering/egg_02TransDecoder
/programs/TransDecoder-v5.5.0/TransDecoder.LongOrfs -t ../egg_01CPC2/egg_noncoding.fasta

/programs/hmmer/bin/hmmsearch --cpu 16 -E 1e-10 --domtblout egg_pfam.domtblout ../../../resources/Pfam-A.hmm egg_noncoding.fasta.transdecoder_dir/longest_orfs.pep

blastp -query egg_noncoding.fasta.transdecoder_dir/longest_orfs.pep -db swissprot -max_target_seqs 1 -outfmt 6 -evalue 1e-5 -out egg_blastp.outfmt6 -remote

/programs/TransDecoder-v5.5.0/TransDecoder.Predict -t ../egg_01CPC2/egg_noncoding.fasta --retain_pfam_hits egg_pfam.domtblout --retain_blastp_hits egg_blastp.outfmt6

grep -o '>MSTRG\S*' egg_noncoding.fasta.transdecoder.cds | grep -o 'MSTRG.[0123456789]*.[0123456789]*' > egg_ORFs.txt

grep -h -v -f egg_ORFs.txt ../egg_01CPC2/egg_noncoding.gtf > egg_nonORFs.gtf
cd ../../..

/programs/gffread-0.9.12/gffread/gffread data/03_filtering/egg_02TransDecoder/egg_nonORFs.gtf -g resources/galaxy_bari06.fa -w data/03_filtering/egg_02TransDecoder/egg_nonORFs.fasta
```

```{bash infernal}
export PATH=/programs/infernal-1.1.3/bin:$PATH
cd resources/rfam

cmscan -o ../../data/03_filtering/egg_03RFAM/infernal.log --noali --tblout ../../data/03_filtering/egg_03RFAM/egg_rfam.tblout --rfam --cpu 24 Rfam.cm ../../data/03_filtering/egg_02TransDecoder/egg_nonORFs.fasta
cd ../..

grep '!' data/03_filtering/egg_03RFAM/egg_rfam.tblout > data/03_filtering/egg_03RFAM/egg_rfam_sig.txt

grep 'sno' data/03_filtering/egg_03RFAM/egg_rfam_sig.txt | grep -o 'MSTRG.[0123456789]*.[0123456789]*'| uniq > data/03_filtering/egg_03RFAM/egg_snoRNA.txt

grep 'ribosomal RNA' data/03_filtering/egg_03RFAM/egg_rfam_sig.txt | grep -o 'MSTRG.[0123456789]*.[0123456789]*'| uniq > data/03_filtering/egg_03RFAM/egg_rRNA.txt

grep 'microRNA' data/03_filtering/egg_03RFAM/egg_rfam_sig.txt | grep -o 'MSTRG.[0123456789]*.[0123456789]*'| uniq > data/03_filtering/egg_03RFAM/egg_miRNA.txt

grep 'tRNA' data/03_filtering/egg_03RFAM/egg_rfam_sig.txt | grep -o 'MSTRG.[0123456789]*.[0123456789]*'| uniq > data/03_filtering/egg_03RFAM/egg_tRNA.txt

grep -o 'MSTRG.[0123456789]*.[0123456789]*' data/03_filtering/egg_03RFAM/egg_rfam_sig.txt | uniq > data/03_filtering/egg_03RFAM/egg_knownRNA.txt

grep -h -v -f data/03_filtering/egg_03RFAM/egg_knownRNA.txt data/03_filtering/egg_02TransDecoder/egg_nonORFs.gtf > data/03_filtering/egg_03RFAM/egg_unknownRNA.gtf
```

```{bash feature-counts}
/programs/subread-2.1.1/bin/featureCounts -T 24 -p -a data/03_filtering/egg_03RFAM/egg_unknownRNA.gtf -F 'GTF' -t transcript -o data/03_filtering/egg_04FeatureCounts/egg_counts.txt data/02_transcripts/egg_01mapped/*.bam

grep 'MSTRG' data/03_filtering/egg_04FeatureCounts/egg_counts.txt > data/03_filtering/egg_04FeatureCounts/egg_counts_no_metadata.txt
```

```{r filter-low-expression}
library(dplyr)
egg_counts_no_metadata <- read.csv("data/03_filtering/egg_04FeatureCounts/egg_counts_no_metadata.txt", sep = "\t", header = FALSE)

expression_level <- function(min_reads, min_count, arr){
  sum_reads <- 0
  for (num in arr){
    if (num >= min_count){
      sum_reads <- sum_reads + 1
    }
  }
  return (sum_reads >= min_reads)
}

rows_to_keep <-c()
for (r in 1:nrow(egg_counts_no_metadata)){
  if(expression_level(3,3,egg_counts_no_metadata[r,7:108])){
    rows_to_keep<- append(rows_to_keep,r)
  }
}

egg_counts_filtered <- egg_counts_no_metadata[rows_to_keep,]
write.table(egg_counts_filtered$V1, file = "data/03_filtering/egg_04FeatureCounts/egg_lncRNA.txt", col.names = FALSE, row.names=FALSE, sep="\t", quote = FALSE)
```
